mettle/worker/worker.go

```diff
@@ -171,23 +171,34 @@ func RunOne(instanceName, name, storage, dir, cacheDir, sandbox, altSandbox, tok
 	return nil
 }
 
-func runForever(instanceName, requestQueue, responseQueue, name, storage, dir, cacheDir, browserURL, sandbox, altSandbox, lucidity, promGatewayURL, tokenFile, redis, redisPassword string, redisTLS bool, cachePrefix, cacheParts []string, clean, secureStorage bool, maxCacheSize, minDiskSpace int64, memoryThreshold float64, versionFile string, costs map[string]mettlecli.Currency, ackExtension time.Duration) error {
+func runForever(instanceName, requestQueue, responseQueue, name, storage, dir, cacheDir, browserURL, sandbox, altSandbox, lucidity, promGatewayURL, tokenFile, redis, redisPassword string, redisTLS bool, cachePrefix, cacheParts []string, clean, secureStorage bool, maxCacheSize, minDiskSpace int64, memoryThreshold float64, versionFile string, costs map[string]mettlecli.Currency, ackExtension time.Duration, immediateShutdown bool) error {
 	w, err := initialiseWorker(instanceName, requestQueue, responseQueue, name, storage, dir, cacheDir, browserURL, sandbox, altSandbox, lucidity, promGatewayURL, tokenFile, redis, redisPassword, redisTLS, cachePrefix, cacheParts, clean, secureStorage, maxCacheSize, minDiskSpace, memoryThreshold, versionFile, costs, ackExtension)
 	if err != nil {
 		return err
 	}
 	defer w.ShutdownQueues()
-	ctx, cancel := context.WithCancel(context.Background())
 	ch := make(chan os.Signal, 2)
 	signal.Notify(ch, syscall.SIGHUP, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGABRT, syscall.SIGTERM)
+	shutdownCh := make(chan bool, 1)
+	defer close(shutdownCh)
 	go func() {
 		sig := <-ch
-		log.Warning("Received signal %s, shutting down when ready...", sig)
-		cancel()
-		w.Report(false, false, false, "Received signal %s, shutting down when ready...", sig)
-		sig = <-ch
-		w.Report(false, false, false, "Received another signal %s, shutting down immediately...", sig)
-		log.Fatalf("Received another signal %s, shutting down immediately", sig)
+		if immediateShutdown || w.actionDigest == nil {
+			w.forceShutdown(fmt.Sprintf("Received shutdown signal %s, shutting down...", sig))
+		}
+		log.Warning("Received signal %s, shutting down when the task completes or times out", sig)
+		shutdownCh <- true
+
+		t := time.NewTicker(5 * time.Minute)
```

Comment:
I don't think you need a ticker here? IIUC it only happens once so this could be
```
select {
case <-time.After(5 * time.Minute):
     ....
case sig := <-ch:
    ....
}
```